gossip/src/duplicate_shred_handler.rs

```diff
@@ -139,21 +172,31 @@ impl DuplicateShredHandler {
                 return false;
             }
         }
-        // Also skip frozen slots or slots with an older proof than me.
+        // Also skip frozen slots or slots with a different proof than me.
         match self.chunk_map.get(&slot) {
-            Some(SlotStatus::Frozen) => {
-                return false;
-            }
-            Some(SlotStatus::UnfinishedProof(slot_map)) => {
-                if let Some(proof_chunkmap) = slot_map.peek(&data.from) {
-                    if proof_chunkmap.wallclock < data.wallclock {
-                        return false;
-                    }
-                }
+            Some(SlotStatus::Frozen) => false,
+            Some(SlotStatus::UnfinishedProof(slot_map)) => match slot_map.get(&data.from) {
+                None => true,
+                Some(proof_chunkmap) => proof_chunkmap.wallclock == data.wallclock,
+            },
+            None => true,
+        }
+    }
+
+    fn dump_pubkeys_with_low_stakes(
+        cached_staked_nodes: Arc<HashMap<Pubkey, u64>>,
+        slot_chunk_map: &mut SlotChunkMap,
+    ) {
+        let mut heap = BinaryHeap::new();
+        for oldkey in slot_chunk_map.keys() {
+            let oldkey_stake = cached_staked_nodes.get(oldkey).copied().unwrap_or_default();
+            heap.push(Reverse::<(u64, Pubkey)>((oldkey_stake, *oldkey)));
+        }
+        for _ in 0..(MAX_PUBKEY_PER_SLOT / 2) {
+            if let Some(Reverse::<(u64, Pubkey)>((_, key))) = heap.pop() {
+                slot_chunk_map.remove(&key);
```

Comment:
Heap would result in an `n log n` performance here.
You can get a linear performance by using `select_nth_unstable`:
https://github.com/solana-labs/solana/blob/fd3f26380/gossip/src/crds.rs#L620-L627

Reply:
I originally chose heap because I read that worst case performance of select_nth_unstable is O(n^2), but you are right, first of all we probably won't hit worst case because pubkey incoming order is pretty random, second select_nth_unstable is clearer for understanding, and for capacity of ~300, O(n^2) or O(nlog(n)) doesn't make much difference.

Changed